src/components/app/index.tsx

```diff
@@ -105,38 +136,53 @@ export default class App extends Component<Props, State> {
   }
 
   async updateImage(index: number) {
-    const { sourceData, images } = this.state;
+    const { sourceData, sourceFile, images } = this.state;
     if (!sourceData) return;
     let image = images[index];
     // Each time we trigger an async encode, the ID changes.
     const id = ++image.counter;
     image.loading = true;
     this.setState({ });
-    let result = await this.updateCompressedImage(sourceData, image.type, image.options);
+    if (image.type === 'original' && sourceFile) {
+      image.extension = sourceFile.name.split('.').pop() || '';
+      // If we don't use this then image data is a bitmap from identity encoder.
+      image.downloadUrl = URL.createObjectURL(new Blob([sourceFile]));
+    }
+    const compressedImage = await this.updateCompressedImage(sourceData, image.type, image.options);
+    const imageBitmap = await createImageBitmap(compressedImage.data);
     image = this.state.images[index];
     // If another encode has been intiated since we started, ignore this one.
     if (image.counter !== id) return;
-    image.data = result;
+    image.data = imageBitmap;
+    if (image.type !== 'original') {
+      if (compressedImage.data instanceof Blob) {
+        image.downloadUrl = URL.createObjectURL(compressedImage.data);
+      } else {
+        image.downloadUrl = URL.createObjectURL(new Blob([compressedImage.data.data]));
+      }
+    }
     image.loading = false;
     this.setState({ });
   }
 
-  async updateCompressedImage(sourceData: ImageData, type: ImageType, options: CodecOptions) {
+  async updateCompressedImage(sourceData: ImageData, type: ImageType, options: CodecOptions): Promise<CompressedImage> {
+    let imageInfo: CompressedImage = { type: '', data: new Blob() };
```

Comment:
Q: what about creating a Blob with the correct mimetype here to avoid the wrapping object, then checking it as .type elsewhere?

Reply:
iirc I might not have had access to the mimetype at this point... need to check.

Reply:
You do on master. Each encoder has a mime type and extension associated with it. Except for identity, where you just reuse the file source.

I set this up in the options branch in anticipation of this. You're welcome. Five peer bonuses please.

Reply:
(the options branch is now on master)

Reply:
sg - going through it now. I expect to close this PR and replace with another.